src/v/rpc/connection_cache.cc

```diff
@@ -99,4 +191,148 @@ ss::future<> connection_cache::stop() {
+  connection_allocation_strategy::changes changes,
+  std::optional<connection_config> config) {
+    auto units = co_await _coordinator_state->mtx.get_units();
+
+    // Add connections we have the config for.
+    if (config) {
+        for (auto& con_add : changes.add_connections) {
+            if (con_add.node == config->dest_node) {
+                co_await container().invoke_on(
+                  con_add.shard,
+                  [cfg = *config](connection_cache& cache) mutable {
+                      if (cache.is_shutting_down()) {
+                          return ss::now();
+                      }
+                      return cache._cache.try_add_or_update(
+                        cfg.dest_node,
+                        std::move(cfg.addr),
+                        std::move(cfg.tls_config),
+                        std::move(cfg.backoff));
+                  });
+            }
+        }
+    }
+
+    // Add shard mappings
+    co_await ss::parallel_for_each(changes.add_mapping, [this](auto map) {
+        return add_or_update_connection_location(
+          map.src_shard, map.node, map.conn_shard);
+    });
+
+    // Remove any shard mappings
+    co_await ss::parallel_for_each(changes.remove_mapping, [this](auto map) {
+        return remove_connection_location(map.shard, map.node);
+    });
+
+    // Remove any connections
+    co_await ss::parallel_for_each(
+      changes.remove_connections, [this](auto con) {
+          return container().invoke_on(con.shard, [con](auto& cache) {
+              if (cache.is_shutting_down()) {
+                  return ss::now();
+              }
+              return cache._cache.remove(con.node);
+          });
+      });
+}
+
+ss::future<> connection_cache::remove_broker_client_coordinator(
+  model::node_id self, model::node_id dest) {
+    vassert(ss::this_shard_id() == _coordinator_shard, "not the coordinator");
+
+    if (is_shutting_down()) {
+        co_return;
+    }
+    auto holder = _gate.hold();
+
+    auto& alloc_strat = _coordinator_state->alloc_strat;
+    if (alloc_strat.has_connection_assignments_for(dest)) {
+        auto changes = alloc_strat.remove_connection_assignments_for(dest);
+        co_await apply_changes(changes, std::nullopt);
+    }
+}
+
+ss::future<>
+connection_cache::remove_broker_client(model::node_id self, model::node_id id) {
+    if (is_shutting_down()) {
+        co_return;
+    }
+    auto holder = _gate.hold();
+
+    co_await container().invoke_on(
+      _coordinator_shard, [self, id](auto& coordinator) mutable {
+          return coordinator.remove_broker_client_coordinator(self, id);
+      });
+}
+
+ss::future<>
+connection_cache::update_broker_client_coordinator(connection_config cfg) {
+    vassert(ss::this_shard_id() == _coordinator_shard, "not the coordinator");
+
+    if (is_shutting_down()) {
+        co_return;
+    }
+    auto holder = _gate.hold();
+
+    auto& alloc_strat = _coordinator_state->alloc_strat;
```

Comment:
Potenial access to _coordinator_state->alloc_strat accessed without owning the mutex?